src/protocol/context/malicious.rs

```diff
@@ -53,21 +54,54 @@ impl<'a, F: Field> MaliciousContext<'a, F> {
         record_id: RecordId,
         input: Replicated<F>,
     ) -> Result<MaliciousReplicated<F>, Error> {
-        self.inner.upgrade(record_id, input).await
+        self.upgrade_sparse(record_id, input, ZeroPositions::Pvvv)
+            .await
+    }
+
+    /// Upgrade a sparse input using this context.
+    /// # Errors
+    /// When the multiplication fails. This does not include additive attacks
+    /// by other helpers.  These are caught later.
+    pub async fn upgrade_sparse(
+        &self,
+        record_id: RecordId,
+        input: Replicated<F>,
+        zeros_at: ZeroPositions,
+    ) -> Result<MaliciousReplicated<F>, Error> {
+        self.inner.upgrade(record_id, input, zeros_at).await
+    }
+
+    /// Upgrade an input for a specific bit index using this context.  Use this for
+    /// inputs that have multiple bit positions in place of `upgrade()`.
+    /// # Errors
+    /// When the multiplication fails. This does not include additive attacks
+    /// by other helpers.  These are caught later.
+    pub async fn upgrade_with<SS: Substep>(
```

Comment:
I changed how the per-bit upgrade works here.  Now you will need to provide a `BitOpStep` if you want to upgrade multiple bits at once.  This is more flexible in that you can upgrade other things.  That turned out to be necessary (see the upgrades in the sparse tests).